docs/_docs/02_testing-best-practices/testing-plan-output.md

```diff
@@ -0,0 +1,40 @@
+---
+layout: collection-browser-doc
+title: Testing plan output
+category: testing-best-practices
+excerpt: >-
+  Testing specific changes described in Terraform plan output
+tags: ["testing-best-practices"]
+order: 202
+nav_title: Documentation
+nav_title_link: /docs/
+---
+
+Sometimes you can get the information you need in your tests from a `terraform plan`; you may not need to do an `apply`.  Terratest provides a way to use certain information from a `terraform plan` in your tests.
+
+The `InitAndPlanWithInfo` function will return a `PlanInfo` struct that will contain information about which resources would be changed and what specifically about them would change.
+
+For example, let's say you have some `web_server` resources that you're creating with [`for_each`](https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings), and you're using the `for_each` values in the name:
+
+```hcl
+resource "aws_instance" "web_server" {
+  for_each = ["env1", "env2", "env3"]
+  name = "web_server_${each.value}"
+  ...
+}
+
+if you want to assert that the names of `web_server` resources that are definied with a terraform `for_each`, you can test this by making assertions against info in this `PlanInfo` struct:
+
+```go
+plan := terraform.InitAndPlanWithInfo(t, terraformOptions)
+
+for _, rc := range plan.PlannedValues {
+    if rc.Name == "web_server" {
+        expected := fmt.Sprintf("web_server_%s", rc.Index)
+        assert.Equal(rc.Values["name"], expected)
+    }
```

Comment:
NIT: You would probably also want to make sure that you reach this block. E.g this test will actually pass if `plan.PlannedValues` was empty, which is probably not what you intended.

Reply:
Good call, updated.  This is where I think some helper methods on the PlanInfo struct could be helful,  stuff like
```
func (p *PlanInfo) GetResourcesByName(name string)...
func (p *PlanInfo) GetUpdatedResources()...
func (p *PlanInfo) GetResourcesByType(type string)...
etc
```

Reply:
Makes sense given that gnarly json!